src/modules/src/controller_indi.c

```diff
@@ -0,0 +1,311 @@
+
+  // The inputs are scaled in order to avoid overflows
+  float du[3];
+  float_vect_copy(du, est->u_d, 3);
+  float_vect_scale(du, INDI_EST_SCALE, 3);
+  est->g1.p = est->g1.p - (est->g1.p * du[0] - est->rate_dd[0]) * du[0] * est->mu;
+  est->g1.q = est->g1.q - (est->g1.q * du[1] - est->rate_dd[1]) * du[1] * est->mu;
+  float ddu = est->u_dd[2] * INDI_EST_SCALE / PERIODIC_FREQUENCY;
+  float error = (est->g1.r * du[2] + est->g2 * ddu - est->rate_dd[2]);
+  est->g1.r = est->g1.r - error * du[2] * est->mu / 3.0f;
+  est->g2 = est->g2 - error * 1000.0f * ddu * est->mu / 3.0f;
+
+  //the g values should be larger than zero, otherwise there is positive feedback, the command will go to max and there is nothing to learn anymore...
+  if (est->g1.p < 0.01f) { est->g1.p = 0.01f; }
+  if (est->g1.q < 0.01f) { est->g1.q = 0.01f; }
+  if (est->g1.r < 0.01f) { est->g1.r = 0.01f; }
+  if (est->g2   < 0.01f) { est->g2 = 0.01f; }
+
+  if (indi.adaptive) {
+    //Commit the estimated G values and apply the scaling
+    indi.g1.p = est->g1.p * INDI_EST_SCALE;
+    indi.g1.q = est->g1.q * INDI_EST_SCALE;
+    indi.g1.r = est->g1.r * INDI_EST_SCALE;
+    indi.g2   = est->g2 * INDI_EST_SCALE;
+  }
+}
+
+void controllerINDIInit(void)
+{
+	float_rates_zero(&indi.angular_accel_ref);
+	float_rates_zero(&indi.u_act_dyn);
+	float_rates_zero(&indi.u_in);
+
+	// Re-initialize filters
+	indi_init_filters();
+}
+
+bool controllerINDITest(void)
+{
+	return true;
+}
+
+void controllerINDI(control_t *control, setpoint_t *setpoint,
+                                         const sensorData_t *sensors,
+                                         const state_t *state,
+                                         const uint32_t tick)
+{
+
+	/*
+	 * 1 - Update the gyro filter with the new measurements.
+	 */
+
+	float stateAttitudeRateRoll = radians(sensors->gyro.x);
+	float stateAttitudeRatePitch = -radians(sensors->gyro.y);
+	float stateAttitudeRateYaw = radians(sensors->gyro.z);
+
+	 struct FloatRates body_rates = {
+			  .p = stateAttitudeRateRoll,
+			  .q = stateAttitudeRatePitch,
+			  .r = stateAttitudeRateYaw,
+	 };
+	 filter_pqr(indi.rate, &body_rates);
+
+
+	/*
+	 * 2 - Calculate the derivative with finite difference.
+	 */
+
+	 finite_difference_from_filter(indi.rate_d, indi.rate);
+
+
+	/*
+	 * 3 - same filter on the actuators (or control_t values), using the commands from the previous timestep.
+	 */
+	 filter_pqr(indi.u, &indi.u_act_dyn);
+
+
+	/*
+	 * 4 - Calculate the desired angular acceleration by:
+	 * 4.1 - Rate_reference = P * attitude_error, where attitude error can be calculated with your favorite
+	 * algorithm. You may even use a function that is already there, such as attitudeControllerCorrectAttitudePID(),
+	 * though this will be inaccurate for large attitude errors, but it will be ok for now.
+	 * 4.2 Angular_acceleration_reference = D * (rate_reference – rate_measurement)
+	 */
+	 float referenceAttitudeRateRoll = radians(setpoint->attitudeRate.roll);
```

Comment:
Is the desired roll rate already calculated somewhere else?

Reply:
I think so yes. Both PID and Mellinger controllers use this value like that. Check it and see if you agree:
- https://github.com/amreelab/crazyflie-firmware/blob/d1fff167ce73647be9e93d2196a001e1d27816c5/src/modules/src/controller_mellinger.c#L254
- https://github.com/amreelab/crazyflie-firmware/blob/d1fff167ce73647be9e93d2196a001e1d27816c5/src/modules/src/controller_pid.c#L64

Reply:
Ok, there should be a gain on the attitude error that produces the desired rate, it would be good if you can find that such that we can change its value if needed.

Reply:
Maybe here?

https://github.com/amreelab/crazyflie-firmware/blob/d1fff167ce73647be9e93d2196a001e1d27816c5/src/modules/src/attitude_pid_controller.c#L115

Reply:
If we follow on this path, I believe we end up with the gains declared in `pid.h`:
https://github.com/amreelab/crazyflie-firmware/blob/d1fff167ce73647be9e93d2196a001e1d27816c5/src/modules/interface/pid.h#L33

Reply:
But do we need to call a PID function that uses these gains, or is that already done outside of our INDI controller file? If you are comparing with the gains that I gave, take into account that I was using radians and quaternions, while here sometimes I see radians and sometimes degrees.
If you'll use this, consider that we don't need integrator gains as INDI is already sort of an integrator. I also think it doesn't make sense to have overlapping gains (e.g. D gain for the attitude as well as a P gain for the rates).

Reply:
As far as I understand, it is already done outside of our INDI controller. As the Mellinger controller only use radians as well, and we are using these values as they come in the same way the Mellinger is using, I think we have a tested proof that our approach works regarding units (radians x degrees). Take a look at how Mellinger does here:
https://github.com/amreelab/crazyflie-firmware/blob/d1fff167ce73647be9e93d2196a001e1d27816c5/src/modules/src/controller_mellinger.c#L250
Maybe what you are trying to tell me is that we will not be able to use these setpoints as they come from the PID rate controller, but we will have to recalculate them. Is that so?